### ISSUE TITLE
[RFC]: Refactor tool parsers to eliminate coding errors and allow more efficient implementations.

### ISSUE THREAD

**Issue by elementary-particle** [vllm-project/vllm#11522]
### Motivation.

Currently the tool parsers are buggy when used and are messy in terms of code, especially in the implementations of `extract_tool_calls_streaming`. Moreover, in the long term, maintaining the entire output string in the chat streaming server and parsing the entire output over and over again for each generated token will become very expensive. This will soon become a performance bottleneck in long tool calls.

Most of the implemented tool parsers aren't carefully written either in terms of correctness nor in terms of efficiency causing a lot of issues already in this repository. A complete refactor of this part of the frontend will be required sooner of later.

So now is probably the best opportunity to refactor things before more tool calling support is added.
From the architectural perspective, clearly it's should be the tool parser's job to maintain states that it needs, so if they need the complete output, they should maintain them with `delta_text` and `delta_token_ids` in their class instead of relying on the server. On the other hand, this encourages the tool parser to use custom, efficient alternative implementations to generate delta messages based on e.g. a internal JSON parser state without needing to parse the full output string every time anymore.
No capability is lost. Performance gains are good and this must be done sooner or later anyway.

No changes to user interface is required and only a few lines of modification is required for the chat server (Only remove the output caching part). The major refactoring should happen in the tool parsers. However, as stated above, we can just move the string cache into the parsers first and refactor them gradually.

### Proposed Change.

This RFC is actually pretty simple so it barely falls into the criteria of 'Large Changes' but I still propose it here.

- [ ] Add internal output cache to existing tool parsers and modify tool parser interface.
- [ ] Fix bugs/refactor parts of the parsing logic in existing tool parsers.
- [ ] Add appropriate tests for the tool parsers.

### Feedback Period.

This shouldn't be hard so 1-2 weeks.

### CC List.

@mgoin @simon-mo @WoosukKwon @petersalas @comaniac @SolitaryThinker @ywang96 @DarkLight1337

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by elementary-particle** [comment]
@cedonley Maybe you can provide some insights on this?

**Comment by cedonley** [comment]
Adding @K-Mistele as well.

I’ve only looked at the streaming implementation, but I agree that coordination between the server and the tool parsers in the current implementation makes tool parsing very fragile. Also agree that refactoring would offer opportunities to optimize.

That said, there are enough commonalities between how the parsing happens that having the tool parsers extend a BaseToolParser that provides a basic state tracking and tool token parsing would be helpful in preventing each parser from being a free-for-all. In fact, this is already happening — if you look at the llama, mistral, and qwen parsers, there are whole gobs of code that are practically copy and paste, but with slight variations. When I was fixing one, I had to make practically the same fix in all the others.

On the parsers themselves, I think we should remove all of the JSON loads/dumps from the parsers. Ideally we have access to the locations in the raw tool string where each start/end token occurs and the position in the string we’ve already returned. Then you just need to return the delta from whatever that is. This also allows us to remove most of the index/rindex and regex that happens inside the tool parser streaming loop.

Agree that this isn’t a huge project. The real work beyond design is just refactoring all of the existing parsers and testing them…but given that the number of future parsers is likely to go up, I agree it’s better to do this before things get completely out of hand…

**Comment by elementary-particle** [comment]
I've written a draft for this idea refactoring the hermes parser. Take a look at #11554. With a streaming json parsers it's actually easier to write correct code and improve performance.

**Comment by elementary-particle** [comment]
As the original author of the tool calling features, @K-Mistele, we definitely want to hear your ideas on this as well.

**Comment by K-Mistele** [comment]
Thanks for opening this! Yes it has become increasingly obvious that a more stable strategy is needed for tool parsing due to the large number of un-anticipated edge & corner cases. 

To my mind, the right approach seems to be to use a finite state machine to model streaming parsing, since it's a very discrete and easy-to-model process; this keeps things relatively atomic and avoids the spaghetti code, and also provides more straightforward ways to manage failure modes. 

I am not familiar with what's available for this in the python ecosystem, but something like xstate (JS ecosystem) probably be a good solution. Here's how I would break it down:

### States
1. generating text 
2. generating function name
3. generating function arguments
4. tool call complete
5. Error state - when an event happens that is not supported for the current state, this should be targeted and should result in a defined error mode e.g. an error delta (not sure if openAI supports this? but it's not possible to send a 5XX once streaming has already started)
6. (additional states pertinent to the model, e.g. generating reasoning tokens)

### Events + transitions:
1. **tool call start** - defined on "generating text", targets "generating function name"; defined on "tool call completed" and targets "generating function name" for parallel tool calls; on other states should result in an error mode
2. **tool call name generated** - defined on "generating function function name", targets "generating function arguments"
3. **tool call arguments completed** - defined on "Generating function arguments" for when the generated arguments are fully JSON-parsable without JSON autcompletion, targets "tool call completed"

Probably there needs to be a way to support transitioning from "Tool call completed" back to "Generating text" since some models will interleave tool calls and text generation e.g. the hermes models, possibly other OSS reasoning models. 

Curious to hear your feedback on this - it's been floating around in my head for a while and would love to collaborate on this if it seems like a good idea. 

cc @mgoin

**Comment by elementary-particle** [comment]
> That said, there are enough commonalities between how the parsing happens that having the tool parsers extend a BaseToolParser that provides a basic state tracking and tool token parsing would be helpful in preventing each parser from being a free-for-all.

Indeed, I think we could construct two types of basic tool parsers:

1. A parser that accepts text before a BOT token and followed by a JSON tool call. More BOT tokens & tool calls could follow in parallel calls.
2. A parser that parses JSON wrapped by BOT tokens and EOT tokens as tool calls, and taking the rest as text response.

> To my mind, the right approach seems to be to use a finite state machine to model streaming parsing, since it's a very discrete and easy-to-model process; this keeps things relatively atomic and avoids the spaghetti code, and also provides more straightforward ways to manage failure modes.

I agree, we need a reliable way to record the required states of the parser. In some cases however the language of the tool call might not be one that can be handled by a FSM(or equivalently, a regular expression), e.g. JSON. I think it's valid in these cases to use a more complex, potentially external library based parser. Most of the tool calling languages probably fall into LL(k) and can be handled with ease.

That said, I think it's a good idea to use a well tested stream parsing library like `ijson` to avoid hand writing some of this code.

**Comment by lulmer** [comment]
Hello everyone, 
I’m very interested in this RFC as I’ve also noticed somewhat slow performance on an in-house tool called “Benchmark” that I’m running. Occasionally, I observe that some calls take around 1200 ms to complete (using LLaMA 3.3-70B on 4 H100 GPUs), while other calls unexpectedly take over 100,000 ms to finish. From a non-streaming perspective, I don’t fully understand how we can avoid using regex to accommodate different tool response formats.

Additionally, I’d like to propose a new parser for LLaMA that improves the accuracy of tool calls. However, it’s currently just a “copy-paste” of the existing design and would likely not include the cool stuff you are mentioning about FSM etc. Should I open a PR anyway?

**Comment by K-Mistele** [comment]
For anyone on this thread, a PR was just merged for extracting reasoning content from deepseek & similar models. Worth discussing here, as I think we will want to move towards providing a unified "model parser" interface that provides tool parsing and/or reasoning parsing depending on what the model supports. FSMs definitely seem like they would be the way to go here, due to the complexities of managing state. 

pinging @interstellarninja here too since the Nous team's Hermes model is capable of both deepseek-style XML-delimited chain-of-thought and XML-delimited tool calls

**Comment by interstellarninja** [comment]
currently Hermes reasoning inside xml tags can be returned as part of text content while tool calls get parsed

but ideally it would be great to have the parsing support for reasoning as well

@mudler from localAI had experimented with this kind of support Hermes a while back

**Comment by gaocegege** [comment]
I'm curious about how FSMs handle JSON, given its flexibility. This is also why structured output engines like xgrammar convert JSON to BNF and implement various optimizations to enhance performance.
